app/src/main/java/org/p2p/wallet/settings/ui/settings/adapter/NewSettingsAdapter.kt

```diff
@@ -1,62 +1,94 @@
 package org.p2p.wallet.settings.ui.settings.adapter
 
+import androidx.recyclerview.widget.DiffUtil
 import androidx.recyclerview.widget.RecyclerView
 import androidx.viewbinding.ViewBinding
 import android.view.ViewGroup
 import org.p2p.wallet.R
-import org.p2p.wallet.settings.ui.settings.presenter.SettingsItem
-import org.p2p.wallet.settings.ui.settings.adapter.viewholder.ComplexSettingViewHolder
+import org.p2p.wallet.settings.model.SettingsItem
+import org.p2p.wallet.settings.model.SettingsItem.ComplexSettingsItem
+import org.p2p.wallet.settings.model.SettingsItem.SettingsGroupTitleItem
+import org.p2p.wallet.settings.model.SettingsItem.SettingsSpaceSeparatorItem
+import org.p2p.wallet.settings.model.SettingsItem.SignOutButtonItem
+import org.p2p.wallet.settings.model.SettingsItem.SwitchSettingsItem
+import org.p2p.wallet.settings.model.SettingsItem.TextSettingsItem
+import org.p2p.wallet.settings.ui.settings.adapter.viewholder.ComplexSettingsViewHolder
+import org.p2p.wallet.settings.ui.settings.adapter.viewholder.NewSettingsViewHolder
 import org.p2p.wallet.settings.ui.settings.adapter.viewholder.SettingsGroupTitleViewHolder
 import org.p2p.wallet.settings.ui.settings.adapter.viewholder.SettingsSpaceSeparatorViewHolder
-import org.p2p.wallet.settings.ui.settings.adapter.viewholder.NewSettingsViewHolder
 import org.p2p.wallet.settings.ui.settings.adapter.viewholder.SignOutButtonViewHolder
-import org.p2p.wallet.settings.ui.settings.adapter.viewholder.SwitchSettingViewHolder
-import org.p2p.wallet.settings.ui.settings.adapter.viewholder.TextSettingViewHolder
+import org.p2p.wallet.settings.ui.settings.adapter.viewholder.SwitchSettingsViewHolder
+import org.p2p.wallet.settings.ui.settings.adapter.viewholder.TextSettingsViewHolder
 import org.p2p.wallet.utils.viewbinding.inflateViewBinding
 
 class NewSettingsAdapter(
     private val onSettingsClicked: (SettingsItem) -> Unit
 ) : RecyclerView.Adapter<NewSettingsViewHolder<*, *>>() {
 
-    private val items = mutableListOf<SettingsItem>()
+    private val data = mutableListOf<SettingsItem>()
 
     fun setItems(newItems: List<SettingsItem>) {
-        items.clear()
-        items.addAll(newItems)
-        notifyDataSetChanged()
+        val old = data.toMutableList()
+        data.clear()
+        data.addAll(newItems)
+        DiffUtil.calculateDiff(SettingsItemsDiffCallback(old, data)).dispatchUpdatesTo(this)
+    }
+
+    fun updateSwitchItem(itemId: Int, isSwitched: Boolean) {
+        val itemIndex = data.indexOfFirst { it is SwitchSettingsItem && it.nameRes == itemId }
```

Comment:
im not sure that updating switch item manually is a good idea
we should always fetch the latest actual data from the storage and update the screen from it, so it's one side data flow

Reply:
This method is being updated from the latest storage. 
And this is exceptional case, since the checkbox updates itself. 
And we are updating its state from presenter additionally. 
That's why we need it. 

In general, for sure we should always have one way data flow